build-tools/build.sh

```diff
@@ -0,0 +1,217 @@
+
+git config --global --add safe.directory /github/workspace
+
+
+# ----------------
+# CHECK IF TEMPLATE
+# ----------------
+echo "::group::Checking if this is a template"
+
+template=$(awk -F'=' '/^IS_LIBRARY:=/{print $2}' Makefile)
+if [ "$template" == "1" ]; then
+    echo "is template"
+else
+    echo "is not template"
+fi
+echo "template=$template" >> $GITHUB_OUTPUT
+
+echo "::endgroup::"
+
+
+# ----------------
+# GET PROJECT INFO
+# ----------------
+echo "::group::Getting project info"
+
+if [ "$GITHUB_EVENT_NAME" == "pull_request" ]; then
+    # Fetch the head SHA directly from the PR API
+    API_URL="https://api.github.com/repos/$GITHUB_REPOSITORY/pulls/$GITHUB_PR_NUM"
+    echo "API URL: $API_URL"
+    API_RESPONSE=$(wget -O- --quiet "$API_URL")
+    if [ $? -ne 0 ]; then
+        echo "Error fetching data from GitHub API"
+        exit $?
+    fi
+    echo "API Response: $API_RESPONSE"
+    sha=$(echo "$API_RESPONSE" | jq -r '.head.sha' | head -c 6)
+
+else
+    # Use the commit SHA after the event
+    sha=$(git rev-parse HEAD | head -c 6)
+fi
+
+
+
+function get_sha() {
+    echo "sha=$sha" >> $GITHUB_OUTPUT
+    echo "SHA found: $sha"
+}
+
+function get_version() {
+    version=$(awk -F'=' '/^VERSION:=/{print $2}' Makefile)
+    echo "Version found: $version"
+    echo "version=$version" >> "$GITHUB_OUTPUT"
+    echo $version >> "version.txt"
+}
+
+function get_library_name() { 
+    library_name=$(awk -F'=' '/^LIBNAME:=/{print $2}' Makefile)
+    echo "library_name=$library_name" >> "$GITHUB_OUTPUT"
+    echo "Library name found: $library_name"
+    echo $library_name >> "library_name.txt"
+}
+
+get_sha &
+get_version &
+get_library_name &
+wait
+
+version=$(cat version.txt)
+rm version.txt
+library_name=$(cat library_name.txt)
+rm library_name.txt
+
+echo "Version before setting postfix: $version"
+echo "SHA before setting postfix: $sha"
+
+if [ "$INPUT_NO_COMMIT_HASH" == "true" ]; then
+    postfix="${version}"
+else
+    postfix="${version}+${sha}"
+fi
+echo "Postfix after setting: $postfix"
+echo "postfix=$postfix" >> "$GITHUB_OUTPUT"
+
+name="$library_name@$postfix"
+echo "name=$name" >> "$GITHUB_OUTPUT"
+echo "Name found: $name"
+
+echo "::endgroup::"
+# ----------------
+# BUILDING PROJECT
+# ----------------
+# Pause errors
+set +e
+pros make clean
+ERR_OUTPUT=$(mktemp)
+STD_OUTPUT=$(mktemp)
+# Set IS_LIBRARY to 0 to build the project if $template is 1
+if (($template == 1)); then
+    echo "::group::Building ${name}"
+    echo "Setting IS_LIBRARY to 0"
+    sed -i "s/^IS_LIBRARY:=.*\$/IS_LIBRARY:=0/" Makefile
+fi
+
+
+if [[ "$INPUT_MULTITHREADING" == "true" ]]; then
+    echo "Multithreading is enabled"
+    start_build_time=$SECONDS
+    make quick -j 2> $ERR_OUTPUT | tee $STD_OUTPUT
+    build_time=$((SECONDS - $start_build_time))
+else
+    echo "Multithreading is disabled"
+    start_build_time=$SECONDS
+    make quick 2> $ERR_OUTPUT | tee $STD_OUTPUT
+    build_time=$((SECONDS - $start_build_time))
+fi
```

Comment:
why are build time and start build time handled inside the if statement?

Reply:
Because I wanted to measure just the make portion.

Reply:
>I think VERSION is very useful for other projects. The biggest reason its there is to make it obvious and provide an easy method for the user to check their version, since this is the first troubleshooting step

Alright fixed this

Reply:
> This probably spends more time than it would normally, as the os has to manage the scheduling of the processes. Also why put them in files? Just put them in variables, no?

Idr why I did this, I just set it back to non-multithreaded.

Reply:
the if statement does not take that long, you could also use the `time` command.

Reply:
If you want me to switch to using the time command I can, but this seems to work fine.

Reply:
nah idk how/if the time command would work here. I just don't like the repetition. 
I think it would be more appropriate to have an additional make args variable:
```bash
extra_make_args=""
if [[ "$INPUT_MULTITHREADING" == "true" ]]; then
    echo "Multithreading is enabled"
    extra_make_args="-j";
else
    echo "Multithreading is disabled"
fi

start_build_time=$SECONDS
make quick $extra_make_args 2> $ERR_OUTPUT | tee $STD_OUTPUT
build_time=$((SECONDS - $start_build_time))
```

Reply:
I honestly was considering doing that, but then for some reason didn't :skull:

Reply:
Alright I implemented that.